[UNI-299] Implementation of paramFinder and paramFinder runner

ywcui1990 opened the pull request:
@marionleborgne @vitaly-krugl 

Address the rest of Vitaly's comment for https://github.com/numenta/numenta-apps/pull/571

Summary
* Use Doxygen convention for documentation in param_finder
* Add end-to-end test for param_finder runner
* Add tests for find_parameters, _getModelParams
* Fix many code styling issues

vitaly-krugl commented on unicorn/py/unicorn_backend/param_finder.py:
```diff
@@ -40,7 +39,14 @@
+# Minimum number of rows. If the number of record in the input data is
+# less than MIN_NUM_ROWS, the param_finder will return the default parameters
+MIN_NUM_ROWS = 100
+MIN_ROW_AFTER_AGGREGATION = 1000
```
Document `MIN_ROW_AFTER_AGGREGATION`

ywcui1990 replied:
It is intended to be public. The user may want to use further aggregation that results in a data file with less than 1000 rows.

vitaly-krugl commented on unicorn/py/unicorn_backend/param_finder.py:
```diff
@@ -206,17 +195,22 @@ def find_parameters(timeStamps, values):
+  (timeStamps, values) = data
+  numRecords = len(data[0])
 
   if type(timeStamps[0]) is not datetime.datetime:
```
Pythonic: `if not isinstance(timeStamps[0], datetime.datetime)`

vitaly-krugl commented on unicorn/py/unicorn_backend/param_finder.py:
```diff
@@ -206,17 +195,22 @@ def find_parameters(timeStamps, values):
-  if len(timeStamps) != len(values):
-    raise ValueError('timeStamps and Values must have the same length')
+  (timeStamps, values) = data
+  numRecords = len(data[0])
```
`numRecords = len(timeStamps)`, since you already have `timeStamps`

vitaly-krugl commented on unicorn/py/unicorn_backend/param_finder.py:
```diff
@@ -175,14 +163,15 @@ def _cwt(data, wavelet, widths):
-  :param values: array_like, each entry is a float number
+  @param data (tuple) Contains:
+    'timeStamps' (sequence) Each entry is an datetime.datetime instance
+    'values' (sequence) Each entry is a float number
```
The `data` parameter change here is different from what was suggested in feedback of previous PR and still exhibits the same issues. Here is clarification of previous feedback:

```
def find_parameters(data):
  """
  @param data Sequence of two-tuples (timestamp, scalar), where timestamp is of
      type `datetime.datetime` and scalar is a number (integer or float)
  """
```

For example:
```
result = find_parameters([(datetime(2016, 1, 1)), (2016, 1, 2), ...])
```

vitaly-krugl commented on unicorn/py/unicorn_backend/param_finder.py:
```diff
@@ -206,17 +195,22 @@ def find_parameters(timeStamps, values):
+    return outputInfo
 
   timeStamps = numpy.array(timeStamps, dtype='datetime64[s]')
 
```
In the code block below:

```
  try:
    values = numpy.array(values).astype('float64')
  except:
    raise TypeError('values must be an array with float numbers')
```

Use a more specific exception instead of just `except:`.

ywcui1990 replied:
I changed it to: 
  try:
    values = numpy.array(values).astype("float64")
  except TypeError:
    print "values must be an array with float numbers"

Is it the right way to catch specific exceptions?

vitaly-krugl replied:
No. I replied to this in another comment.

vitaly-krugl commented on unicorn/py/unicorn_backend/param_finder.py:
```diff
@@ -206,17 +195,22 @@ def find_parameters(timeStamps, values):
   """
-  if len(timeStamps) != len(values):
-    raise ValueError('timeStamps and Values must have the same length')
+  (timeStamps, values) = data
```
nit - `timeStamps` doesn't need to be camel-case. `timestamp` is a legitimate word.

vitaly-krugl commented on unicorn/py/unicorn_backend/param_finder.py:
```diff
@@ -231,30 +225,30 @@ def find_parameters(timeStamps, values):
+                                       values,
+                                       medianSamplingInterval)
 
   (_cwtVar, _timeScale) = _calculateContinuousWaveletTransform(
```
Leading underscores are reserved for private members and private module-level constants. Local variables in a function should not have them. These should be `cwtVar, timeScale`

vitaly-krugl commented on unicorn/py/unicorn_backend/param_finder.py:
```diff
@@ -231,30 +225,30 @@ def find_parameters(timeStamps, values):
-    thresh=_AGGREGATION_WINDOW_THRESH,
-    samplingInterval=medianSamplingInterval,
-    numDataPts=numDataPts)
+      timeScale=_timeScale,
```
Per numenta's coding standard, indentation should be two spaces in python code.

vitaly-krugl commented on unicorn/py/unicorn_backend/param_finder.py:
```diff
@@ -231,30 +225,30 @@ def find_parameters(timeStamps, values):
+                           aggFunc),
     "modelInfo": _getModelParams(useTimeOfDay, useDayOfWeek, values),
   }
   return outputInfo
```
nit - this might be more concise without loss of readability:

```
 return {
    "aggInfo": _getAggInfo(medianSamplingInterval,
                           suggestedSamplingInterval,
                           aggFunc),
    "modelInfo": _getModelParams(useTimeOfDay, useDayOfWeek, values),
  }
```

vitaly-krugl commented:
Some of the code in `param_finder.py` is using `@returns` and some is using `@return`. I think that only one of the two is correct.

vitaly-krugl commented on unicorn/py/unicorn_backend/param_finder.py:
```diff
@@ -500,8 +516,7 @@ def _determineEncoderTypes(cwtVar, timeScale):
+      if (timeScale[leftLocalMin] < weekPeriod < timeScale[rightLocalMin]
           and cwtVarAtWeekPeriod > localMaxValue * 0.5):
         useDayOfWeek = True
 
```
In the code snippet

```
 def _getAggregationFunction(medianSamplingInterval,
                             medianAbsoluteDevSamplingInterval,
                             aggregationFuncThresh=0.2):
```

Is the `0.2` default value related to `_AGGREGATION_WINDOW_THRESH` or is it just a coincidence that they have the same value? If they are semantically the same, then this function should use the same named constant instead of the "naked" constant.

Also, does it make sense to have a default value here? It's a private function that is only used once and without passing the `aggregationFuncThresh` arg. Either this parameter is not needed at all (at this time) and `_getAggregationFunction` should simply use a (named) constant internally, or the default value should be removed from the parameter and the caller should pass an explicit value.

vitaly-krugl commented on unicorn/py/unicorn_backend/param_finder.py:
```diff
@@ -519,13 +534,16 @@ def _getAggregationFunction(medianSamplingInterval,
 
```
In the code snippet

```
 def _getAggregationFunction(medianSamplingInterval,
                             medianAbsoluteDevSamplingInterval,
                             aggregationFuncThresh=0.2):
```

Is the `0.2` default value related to `_AGGREGATION_WINDOW_THRESH` or is it just a coincidence that they have the same value? If they are semantically the same, then this function should use the same named constant instead of the "naked" constant.

Also, does it make sense to have a default value here? It's a private function that is only used once and without passing the `aggregationFuncThresh` arg. Either this parameter is not needed at all (at this time) and `_getAggregationFunction` should simply use a (named) constant internally, or the default value should be removed from the parameter and the caller should pass an explicit value.

vitaly-krugl commented on unicorn/py/unicorn_backend/param_finder_runner.py:
```diff
@@ -176,7 +187,7 @@ def _readCSVFile(fileName,
       values.append(float(row[valueIndex]))
 
       numRow += 1
```
numRow => numRows

vitaly-krugl commented:
Need to fix up Sphinx param/return documentation in `param_finder_runner.py`. For example:

```
   :param fileName: str, path to input csv file
   :param rowOffset: int, index of first data row in csv
   :param timestampIndex: int, column index of the timeStamp
   :param valueIndex: int, column index of the value
   :param datetimeFormat: str, datetime format string for python's
                         datetime.strptime
   :return: A two-tuple (timestamps, values), where
            timeStamps is a list of datetime
            values is a list of float data values
```

Should be 
Note: use of param types between keyword param and param's name; also note `:return:` => `:returns:`
```
  :param str fileName: path to input csv file
  :param int rowOffset: index of first data row in csv
  :param int timestampIndex: column index of the timeStamp
  :param int valueIndex: column index of the value
  :param str datetimeFormat: datetime format string for python's
    datetime.strptime
  :returns: A two-tuple (timestamps, values), where
    timeStamps is a list of datetime 
    values is a list of float data values
```

vitaly-krugl commented on unicorn/py/unicorn_backend/param_finder_runner.py:
```diff
@@ -176,7 +187,7 @@ def _readCSVFile(fileName,
       values.append(float(row[valueIndex]))
```
In the code snippet:

```
       timeStamp = datetime.datetime.strptime(row[timestampIndex],
                                              datetimeFormat)
 
       # use local timezone to be consistent with the default assumption
       # in numpy datetime64 object
       timeStamp = timeStamp.replace(tzinfo=tz.tzutc())
```
What if `datetime.datetime.strptime` returned a `datetime` with `tzinfo` already in it? Could the unconditional `replace` have an adverse impact on the resulting parameters in such a scenario?

Also, the comment "use local timezone to be consistent with the default assumption numpy datetime64 object" is not consistent with the code, which now uses UTC time zone (not local).

vitaly-krugl commented on unicorn/tests/py/unit/unicorn_backend/param_finder_runner_test.py:
```diff
@@ -77,37 +76,76 @@ def _assertArgumentPatternPasses(argumentPattern=None):
       datetimeFormat="%Y-%m-%d %H:%M:%S"
     )
+
     self.assertAlmostEqual(values[0], 20.0)
```
Why is "almost equal" needed? Is parsing of `20.0` losing precision?

vitaly-krugl commented on unicorn/tests/py/unit/unicorn_backend/param_finder_runner_test.py:
```diff
@@ -77,37 +76,76 @@ def _assertArgumentPatternPasses(argumentPattern=None):
     )
+
     self.assertAlmostEqual(values[0], 20.0)
     self.assertAlmostEqual(str(timeStamps[0]),
```
Why not simply `self.assertEqual` for the strings? The "almost equal" has some confusing implications for the maintainer/reader/reviewer of the code.

vitaly-krugl commented on unicorn/tests/py/unit/unicorn_backend/param_finder_runner_test.py:
```diff
@@ -77,37 +76,76 @@ def _assertArgumentPatternPasses(argumentPattern=None):
+    csvFd, csvPath = tempfile.mkstemp()
+    self.addCleanup(os.unlink, csvPath)
+
+    with open(csvPath, "wb") as csvFile:
```
Use `os.fdopen` instead so that file descriptor `csvFd` would be closed automatically when scope leaves the context manager; see https://github.com/numenta/numenta-apps/blob/e09e5af00143acf48acd311da160592bf5f099d9/unicorn/tests/py/integration/unicorn_backend/model_runner_2_test.py#L643

vitaly-krugl commented on unicorn/tests/py/unit/unicorn_backend/param_finder_runner_test.py:
```diff
@@ -77,37 +76,76 @@ def _assertArgumentPatternPasses(argumentPattern=None):
+      with patch('sys.stdout', new=StringIO.StringIO()) as fake_out:
+        param_finder_runner.main()
+        outputInfo = fake_out.getvalue()
+        outputInfo = json.loads(outputInfo)
```
nit - a bit more compact without losing readability: `outputInfo = json.loads(fake_out.getvalue())`

vitaly-krugl commented on unicorn/tests/py/unit/unicorn_backend/param_finder_runner_test.py:
```diff
@@ -77,37 +76,76 @@ def _assertArgumentPatternPasses(argumentPattern=None):
+        outputInfo = json.loads(outputInfo)
+
+        # since we used a flat line as test data, there should be no aggregation
+        self.assertTrue(outputInfo["aggInfo"] == None)
```
Use `self.assertIsNone(outputInfo["aggInfo"])` instead. Using the special-purpose assertion methods allows unittest to print more useful diagnostic messages when the assertion fails.

vitaly-krugl commented on unicorn/tests/py/unit/unicorn_backend/param_finder_test.py:
```diff
@@ -172,4 +171,98 @@ def testDetermineAggregationWindow(self):
+    self.assertEqual(aggregationTimeScale, samplingInterval.astype('float64'))
+
+  def testFindParameters(self):
+    def createTestData(type, timeStepSeconds=300):
```
Don't use names that override builtin objects. Here, `type` parameter overrides the builtin `type` function.

vitaly-krugl commented on unicorn/tests/py/unit/unicorn_backend/param_finder_test.py:
```diff
@@ -172,4 +171,98 @@ def testDetermineAggregationWindow(self):
+      return (timeStamps, values)
+
+
+    outputInfo = param_finder.find_parameters(createTestData('flat', 300))
```
In the assertions below, use the special-purpose assertions such as `self.assertIsNone`, `self.assertIsNotNone`, `self.assertGreater`, `self.assertEqual`, etc., to enable unittest to display more useful diagnostic messages when an assertion fails.

For example:

```
    self.assertGreater(outputInfo["aggInfo"]['windowSize'], 300)
    self.assertEqual(outputInfo["aggInfo"]['func'], 'mean')
    # Exclude timeOfDay and dayOfWeek encoder for flat line
    self.assertEqual(outputInfo["modelInfo"]['modelConfig']['modelParams'] \
                      ['sensorParams']['encoders']['c0_timeOfDay'], None)
    self.assertEqual(outputInfo["modelInfo"]['modelConfig']['modelParams'] \
                      ['sensorParams']['encoders']['c0_dayOfWeek'], None)


```

vitaly-krugl commented on unicorn/tests/py/unit/unicorn_backend/param_finder_test.py:
```diff
@@ -172,4 +171,98 @@ def testDetermineAggregationWindow(self):
+
+
+    outputInfo = param_finder.find_parameters(createTestData('flat', 300))
+    self.assertTrue(outputInfo["aggInfo"]['windowSize'] > 300)
```
In numenta-apps, the convention is to use double-quotes

vitaly-krugl commented on unicorn/tests/py/unit/unicorn_backend/param_finder_test.py:
```diff
@@ -172,4 +171,98 @@ def testDetermineAggregationWindow(self):
+
+    outputInfo = param_finder.find_parameters(createTestData('weekly', 7200))
+    self.assertTrue(outputInfo["aggInfo"]['windowSize'] > 7200)
+    self.assertTrue(outputInfo["aggInfo"]['func'] == 'mean')
```
All of these test cases test for "mean". Should there also be a test for "sum"?

vitaly-krugl commented on unicorn/tests/py/unit/unicorn_backend/param_finder_test.py:
```diff
@@ -172,4 +171,98 @@ def testDetermineAggregationWindow(self):
+    self.assertTrue(outputInfo["aggInfo"]['windowSize'] > 7200)
+    self.assertTrue(outputInfo["aggInfo"]['func'] == 'mean')
+    # Use dayOfWeek but not timeOfDay encoder for dataSet with daily period
+    self.assertTrue(outputInfo["modelInfo"]['modelConfig']['modelParams'] \
```
No need for line-continuation backslash. When inside parentheses, python interpreter does the right thing. Our coding convention favors placing long lines in parentheses instead of using the backslash.

vitaly-krugl commented on unicorn/tests/py/unit/unicorn_backend/param_finder_test.py:
```diff
@@ -172,4 +171,98 @@ def testDetermineAggregationWindow(self):
+    values = numpy.linspace(0, 10, 10)
+    modelParams = param_finder._getModelParams(
+      useTimeOfDay=False, useDayOfWeek=False, values=values)
+
```
use `self.assertIsNone`, `self.assertIsNotNone`, etc.

vitaly-krugl commented:
Run pylint, and fix findings. For example, in `param_finder.py`, pylint complains:
```pylint
Line 232	Column 0	C0330: Wrong hanging indentation.
Line 235	Column 0	C0330: Wrong hanging indentation.
Line 236	Column 0	C0330: Wrong hanging indentation.
Line 237	Column 0	C0330: Wrong hanging indentation.
Line 238	Column 0	C0330: Wrong hanging indentation.
Line 239	Column 0	C0330: Wrong hanging indentation.
Line 347	Column 0	C0330: Wrong continued indentation.
Line 348	Column 0	C0330: Wrong continued indentation.
Line 512	Column 0	C0330: Wrong continued indentation.
Line 166	Column 0	C0103: Invalid function name "find_parameters"
Line 478	Column 2	C0103: Invalid variable name "baseline_value"
```

vitaly-krugl commented:
In tests that intentionally access private/protected members, add the following pragma just before the imports to disable such warnings:

```
# Disable pylint warnings concerning access to protected members
# pylint: disable=W0212
```

vitaly-krugl commented:
@ywcui1990, I am done with the review. Please address the findings. Use your judgement regarding feedback tagged as "nit" (for nitpick)

ywcui1990 commented:
@vitaly-krugl Thanks for the detailed feedback. I have addressed your comments and added a new test for transaction data (aggFunc="sum"). I also ran pylint on my code and addressed most of the issues. There are still some remaining pylint issues that seems to be false alert, which are related to (1) order of imports (2) missing members of numpy and (3) unsubscriptable-object issues

vitaly-krugl commented on unicorn/py/unicorn_backend/param_finder.py:
```diff
@@ -206,75 +197,84 @@ def find_parameters(timeStamps, values):
-    raise TypeError('values must be an array with float numbers')
+    values = numpy.array(values).astype("float64")
+  except TypeError:
+    print "values must be an array with float numbers"
```
@ywcui1990, there are a couple of problems with this:

1. The code now suppresses the TypeError exception and subsequent code will be in trouble when it attempts to access the `value` variable.
2. Printing from library code is not a good idea. Such messages go to stdout, and possibly corrupt the application's stdout data stream. So, we definitely don't want to print from here.
3. For now, we could remove this try/except and simply rely on numpy's own message in its exception

vitaly-krugl commented:
Thanks for addressing most of the feedback, @ywcui1990. We're almost there. Please address the new issue with except/print https://github.com/numenta/numenta-apps/pull/602#discussion-diff-52414448

vitaly-krugl commented on unicorn/tests/py/unit/unicorn_backend/param_finder_test.py:
```diff
@@ -169,7 +179,121 @@ def testDetermineAggregationWindow(self):
+                         ["sensorParams"]["encoders"]["c0_dayOfWeek"])
+
+if __name__ == "__main__":
+  unittest.main()
```
add one empty line at end of script per python convention.

ywcui1990 commented:
Thanks @vitaly-krugl , I have addressed the except/print issue. Is there any other issue you noticed?

vitaly-krugl commented:
LGTM

ywcui1990 commented:
Great. Please merge then @vitaly-krugl  I don't have write access to this repository.

vitaly-krugl commented:
Thanks, @ywcui1990
